### ISSUE TITLE
[2.0] Include in rescue block is included but does not run

### ISSUE THREAD

**Issue by dmsimard** [ansible/ansible#13605]
**Issue type**: Bug Report
**Ansible version**: v2.0.0-0.7.rc2
**Ansible configuration**: Default from pip install
**Environment**: Fedora, CentOS, locally reproducible
##### Summary:

Running a command directly in the rescue block works as follows:

```

---
# roles/tasks/main.yml
tasks:
 - block:
     - debug: msg='I execute normally'
     - command: /bin/false
   rescue:
      - debug:
          msg: "rescuing"
```

Output:

```
$ ansible-playbook -vv playbooks/test.yml 
No config file found; using defaults
 [WARNING]: provided hosts list is empty, only localhost is available

1 plays in playbooks/test.yml

PLAY [Include test role] *******************************************************

TASK [test : debug msg=I execute normally] *************************************
ok: [localhost] => {
    "changed": false, 
    "msg": "I execute normally"
}

TASK [test : command] **********************************************************
fatal: [localhost]: FAILED! => {"changed": true, "cmd": ["/bin/false"], "delta": "0:00:00.001340", "end": "2015-12-18 15:10:27.252622", "failed": true, "rc": 1, "start": "2015-12-18 15:10:27.251282", "stderr": "", "stdout": "", "stdout_lines": [], "warnings": []}

TASK [test : debug msg=rescuing] ***********************************************
ok: [localhost] => {
    "changed": false, 
    "msg": "rescuing"
}

PLAY RECAP *********************************************************************
localhost                  : ok=2    changed=0    unreachable=0    failed=1
```

However, if I move that debug to a rescue.yml file and include it instead, the debug does not run. See following example:

```

---
# roles/tasks/main.yml
tasks:
 - block:
     - debug: msg='I execute normally'
     - command: /bin/false
   rescue:
     - include: rescue.yml
```

and

```

---
# roles/tasks/rescue.yml
- debug:
    msg: "rescuing"
```

The output of the ansible run with the include is as follows:

```
$ ansible-playbook -vv playbooks/test.yml 
No config file found; using defaults
 [WARNING]: provided hosts list is empty, only localhost is available

1 plays in playbooks/test.yml

PLAY [Include test role] *******************************************************

TASK [test : debug msg=I execute normally] *************************************
ok: [localhost] => {
    "changed": false, 
    "msg": "I execute normally"
}

TASK [test : command] **********************************************************
fatal: [localhost]: FAILED! => {"changed": true, "cmd": ["/bin/false"], "delta": "0:00:00.001356", "end": "2015-12-18 15:11:35.071738", "failed": true, "rc": 1, "start": "2015-12-18 15:11:35.070382", "stderr": "", "stdout": "", "stdout_lines": [], "warnings": []}

TASK [test : include] **********************************************************
included: /home/dmsimard/dev/ansible/playbooks/roles/test/tasks/rescue.yml for localhost

PLAY RECAP *********************************************************************
localhost                  : ok=2    changed=0    unreachable=0    failed=1
```
##### Expectations:

I would expect an included task file to actually run it's contents, in the examples above, this would mean running the debug module inside the rescue.yml file.

**Comment by dmsimard** [comment]
So I thought this could have been related to https://github.com/ansible/ansible/issues/13378

I tested with the latest tree but it doesn't change the result.

**Comment by jimi-c** [comment]
This fixes the bug, however I have not had time to sufficiently test it:

<pre>
diff --git a/lib/ansible/executor/play_iterator.py b/lib/ansible/executor/play_iterator.py
index 795eed2..d2bb8a3 100644
--- a/lib/ansible/executor/play_iterator.py
+++ b/lib/ansible/executor/play_iterator.py
@@ -397,7 +397,7 @@ class PlayIterator:
 
     def _insert_tasks_into_state(self, state, task_list):
         # if we've failed at all, or if the task list is empty, just return the current state
-        if state.fail_state != self.FAILED_NONE or not task_list:
+        if state.fail_state == self.FAILED_NONE and state.run_state != self.ITERATING_RESCUE or not task_list:
             return state
 
         if state.run_state == self.ITERATING_TASKS:
</pre>

**Comment by dmsimard** [comment]
@jimi-c If you have a branch/tag with this in, I can swap out rc2 with this to give it a try.

**Comment by jimi-c** [comment]
@dmsimard I'm about to commit it to devel/stable-2.0

**Comment by jimi-c** [comment]
Turns out I did have a logic error there, my update patch is:

<pre>
diff --git a/lib/ansible/executor/play_iterator.py b/lib/ansible/executor/play_iterator.py
index d2bb8a3..534f216 100644
--- a/lib/ansible/executor/play_iterator.py
+++ b/lib/ansible/executor/play_iterator.py
@@ -397,7 +397,7 @@ class PlayIterator:
 
     def _insert_tasks_into_state(self, state, task_list):
         # if we've failed at all, or if the task list is empty, just return the current state
-        if state.fail_state == self.FAILED_NONE and state.run_state != self.ITERATING_RESCUE or not task_list:
+        if state.fail_state != self.FAILED_NONE and state.run_state not in (self.ITERATING_RESCUE, self.ITERATING_ALWAYS) or not task_list:
             return state
 
         if state.run_state == self.ITERATING_TASKS:
</pre>

I'll push this into devel for now, and merge to stable-2.0 after a bit more testing has been done.

**Comment by dmsimard** [comment]
So I just tested several roles using includes in rescue blocks (see https://github.com/redhat-openstack/weirdo/commit/5c822cf960c51d0a0c404b80faa6654eeeca0212) and it worked well for me. It didn't break anything in the CI.

**Comment by jimi-c** [comment]
@dmsimard I've merged this into stable-2.0 as well now, so we'll go ahead and close this issue.

If you continue seeing any problems related to this issue, or if you have any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:
- https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
- https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular issue is resolved.

Thank you!
